amazon/ion/ion_py_objects.py

```diff
@@ -0,0 +1,605 @@
+from decimal import Decimal
+from collections.abc import MutableMapping
+
+from amazon.ion.core import IonType, IonEvent
+from amazon.ion.symbols import SymbolToken
+
+
+class IonPyNull_new(object):
+    ion_annotations = ()
+    ion_type = IonType.NULL  # TODO initialized to NULL type first, what's the real type?
+
+    __name__ = 'IonPyNull_new'
+    __qualname__ = 'IonPyNull_new'
+
+    def __init__(self, ion_type=None, value=None, annotations=()):
+        self.ion_type = ion_type
+        self.ion_annotations = annotations
+
+    def __bool__(self):
+        return False
+
+    def _copy(self):
+        args, kwargs = self._to_constructor_args(self)
+        value = self.__class__(*args, **kwargs)
+        value.ion_type = self.ion_type
+        value.ion_annotations = self.ion_annotations
+        return value
+
+    @staticmethod
+    def _to_constructor_args(value):
+        return (None, value,), {}
+
+    @classmethod
+    def from_event(cls, ion_event):
+        if ion_event.value is not None:
+            args, kwargs = cls._to_constructor_args(ion_event.value)
+        else:
+            args, kwargs = (None, None, ()), {}
+        value = cls(*args, **kwargs)
+        value.ion_type = ion_event.ion_type
+        value.ion_annotations = ion_event.annotations
+        return value
+
+    @classmethod
+    def from_value(cls, ion_type, value, annotations=()):
+        if value is None:
+            value = IonPyNull_new()
+        else:
+            args, kwargs = cls._to_constructor_args(value)
+            value = cls(*args, **kwargs)
+        value.ion_type = ion_type
+        value.ion_annotations = annotations
+        return value
+
+    def to_event(self, event_type, field_name=None, in_struct=False, depth=None):
+        value = self
+        if isinstance(self, IonPyNull_new) or self.ion_type.is_container:
+            value = None
+
+        if in_struct:
+            if not isinstance(field_name, SymbolToken):
+                field_name = SymbolToken(field_name, 0 if field_name is None else None)
+        else:
+            field_name = None
+
+        return IonEvent(event_type, ion_type=self.ion_type, value=value, field_name=field_name,
+                        annotations=self.ion_annotations, depth=depth)
+
+
+class IonPyDecimal_new(Decimal):
+    ion_annotations = ()
```

Comment:
Did you mean that why not define ion_annotation in `__init__`? Since the method signature might change later so I kept both `ion_type` and `ion_annotation` defined at the same place.  

Correct me if this doesn't answer your question but since I saw similar comments above, I can provide my experiment results. This is kind of related to @JustinDaleGray's first feedback.

> Consider the use of [__slots__](https://docs.python.org/3/reference/datamodel.html#slots) versus the use of __dicts__ or __weakref__. We can further reduce memory allocations by not creating a dictionary for attributes where possible.
 
I first used \_\_slots\_\_ to limit the attributes we want, and I saw a 5% of _object initialization_ improvement. 

Later on I found another approach to optimize object initialization speed by setting `class attributes` instead of `instance attributes`. `instance attributes` refers to the attributes defined by \_\_init__. I saw a small _object initialization_ improvement as well.

However, Python doesn't allow both `__slot__` and `class attributes` optimizations at once. I chose the class attributes optimization since we will change the method signature later and all values are default to its own Ion type and empty annotations unless users assign specific values to them.

I just switched between these three approaches and re-run the deserialization benchmarking, I only saw a 1-2% differences between them (\_\_slot__ > class variables > instance variables). So any of them should work. Do you have any concern for class variables?

Reply:
So let's be clear on our object model and `class` vs `instance` attributes [this SO post](https://stackoverflow.com/questions/12409714/python-class-members) covers the latter well.

In our case we need them to be an instance attribute because each value has it's own annotations.

In the code here you're setting a `class` attribute for `annotations` to the empty tuple. In your `__new__` you are setting an instance attribute that is effectively shadowing the class attribute. So given that you're always shadowing it I don't understand why it exists. 

My concern is that shadowing a class attribute with an instance attributes seems likely to introduce subtle bugs. I would avoid it, particularly since given that you are always shadowing it, I don't understand how the code I see here is more efficient for having it.

Reply:
For `ion_type` it makes perfect sense. I wonder if the optimizations you observed were actually due to just the benefit of setting the `ion_type` once per class?

Reply:
Okay will add the change in the coming commit.

Reply:
I moved `ion_type` to be a class attribute, for specific type such as List and Sexp, I leave it in the constructor so it won't break any existing use case.